File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.java
import java.util.*;
class Solution {
    public int solution(int n, int k, int[] enemy) {
        PriorityQueue<Integer> hp = new PriorityQueue<>();
        if(enemy.length <= k){
            return enemy.length;
        }

        for(int i = 0 ; i < enemy.length ; i++){
            if(hp.size() < k){
                hp.add(enemy[i]);
            } else {
                hp.add(enemy[i]);
                n -= hp.poll();
                if(n < 0){
                    return i;
                }
            }
        }
        return enemy.length;
    }
}

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.kt
import java.util.*
class Solution {
fun solution(n: Int, k: Int, enemy: IntArray): Int {
    var answer: Int = 0
    val pq = PriorityQueue<Int>(Collections.reverseOrder())
    var nn = n
    var kk = k
    for(i in enemy.indices){
        pq.add(enemy[i])
        nn -= enemy[i]
        if(nn < 0){
            while(nn < 0 && kk > 0 && pq.isNotEmpty()){
                val maxEnemy = pq.poll()
                nn += maxEnemy
                kk--
//                print("nn: $nn, kk: $kk, maxEnemy: $maxEnemy, i: $i, enemy[i]: ${enemy[i]}\n")
            }
            if(nn < 0){
                return i
            }
        }
    }
    return enemy.size
}
}